File: Graph_Traversal/BOJ_13023(ABCDE_dfs).cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
int n, m;
vector<int> map[2001];
bool visited[2001];
int answer=0;

void dfs(int cur, int dep) {
	visited[cur] = true;
	if (dep == 4) {
		answer = 1;
		return;
	}
	for (int i = 0; i < map[cur].size(); i++) {
		int next = map[cur][i];
		if (!visited[next]) {
			dfs(next, dep + 1);
			visited[next] = false;
		}
	}
}

int main(void) {
	int result = 0;
	cin >> n >> m;
	for (int i = 0; i < m; i++) {
		int a, b;
		cin >> a >> b;
		map[a].push_back(b);
		map[b].push_back(a);
	}

	for (int i = 0; i < n; i++) {
		dfs(i, 0);
		visited[i] = false;
		if (answer==1)
			break;
	}
	if (answer!=0) cout << '1';
	else cout << '0';
	return 0;
}

File: Graph_Traversal/BOJ_13023(ABCDE_dfs).py
n,m=map(int,input().split())
visited=[False] *(n)
graph=[[] for _ in range(n)]
def dfs(cur,dep):
    global answer
    visited[cur]=True
    if dep==4:
        answer=True
        return
    for i in graph[cur]:
        if not visited[i]:
            dfs(i,dep+1)
            visited[i]=False

for i in range(m):
    a,b=map(int,input().split())
    graph[a].append(b)
    graph[b].append(a)
answer=False
for i in range(n):
    dfs(i,0)
    visited[i]=False
    if answer:
        break
print(1 if answer else 0)